File: nodejs/recognize_stream.js
const fs = require('fs');
const {createSttClient, checkWavFormat} = require('./common.js');
const {buildStreamingRecognizeCommand} = require('./args.js');
const stream = require('stream');
const wav = require('wav');
const util = require('util');


const argv = buildStreamingRecognizeCommand().parse();

function main() {
    const sttClient = createSttClient();
    const sttStreamingCall = sttClient.StreamingRecognize();

    sttStreamingCall.on('metadata', (metadata) => console.log("Initial response metadata", metadata));
    sttStreamingCall.on('status', (status) => {
        console.log("Call ended, status", status);
        sttClient.close();
    });
    sttStreamingCall.on('error', (error) => console.error("Error", error));
    sttStreamingCall.on('data', (response) => console.log(util.inspect(response, false, null, true)));

    const configMessage = {
        streamingConfig: {
            config: {
                encoding: argv.encoding,
                sampleRateHertz: argv.rate,
                languageCode: argv.languageCode,
                maxAlternatives: argv.maxAlternatives,
                numChannels: argv.numChannels,
                enableAutomaticPunctuation: argv.automaticPunctuation,
                doNotPerformVad: (argv.performVad ? undefined : true),
                vadConfig: (!argv.performVad ? undefined : {
                    silenceDurationThreshold: argv.silenceDurationThreshold
                }),
            },
            interimResultsConfig: {
                enableInterimResults: argv.interimResults,
            },
            singleUtterance: argv.singleUtterance,
        }
    };

    sttStreamingCall.write(configMessage);

    const reader = fs.createReadStream(argv.file);
    const transformStream = new stream.Transform({
        readableObjectMode: true,
        transform(chunk, encoding, callback) {
            callback(null, {audioContent: chunk});
        }
    });

    if (argv.file.endsWith(".wav")) {
        const wavReader = wav.Reader();
        wavReader.on('format', (format) => {
            if (!checkWavFormat(format, argv)) {
                sttStreamingCall.end();
                process.exit(1);
            } else {
                wavReader.pipe(transformStream).pipe(sttStreamingCall);
            }
        });
        reader.pipe(wavReader);
    } else {
        reader.pipe(transformStream).pipe(sttStreamingCall);
    }
}

main();

File: golang/cmd/recognize_stream/recognize_stream.go
package main

import (
	"context"
	"github.com/TinkoffCreditSystems/voicekit-examples/golang/pkg/args"
	"github.com/TinkoffCreditSystems/voicekit-examples/golang/pkg/common"
	sttPb "github.com/TinkoffCreditSystems/voicekit-examples/golang/pkg/tinkoff/cloud/stt/v1"
	"io"
	"os"
	"strings"
	"sync"
)

func main() {
	opts := args.ParseStreamingRecognizeOptions()
	if opts == nil {
		os.Exit(1)
	}
	defer opts.InputFile.Close()

	var dataReader io.Reader
	if strings.HasSuffix(opts.InputFile.Name(), ".wav") {
		reader, err := common.OpenWavFormat(opts.InputFile, *opts.Encoding, *opts.NumChannels, *opts.Rate)
		if err != nil {
			panic(err)
		}
		dataReader = reader
	} else {
		dataReader = opts.InputFile
	}

	client, err := common.NewSttClient()
	if err != nil {
		panic(err)
	}
	defer client.Close()

	// NOTE: in production code you should probably use context.WithCancel, context.WithDeadline or context.WithTimeout
	// instead of context.Background()
	stream, err := client.StreamingRecognize(context.Background())
	if err != nil {
		panic(err)
	}
	reqeust := &sttPb.StreamingRecognizeRequest{
		StreamingRequest: &sttPb.StreamingRecognizeRequest_StreamingConfig{
			StreamingConfig: &sttPb.StreamingRecognitionConfig{
				Config: &sttPb.RecognitionConfig{
					Encoding:                   sttPb.AudioEncoding(sttPb.AudioEncoding_value[*opts.Encoding]),
					SampleRateHertz:            uint32(*opts.Rate),
					LanguageCode:               *opts.LanguageCode,
					MaxAlternatives:            uint32(*opts.MaxAlternatives),
					ProfanityFilter:            false,
					EnableAutomaticPunctuation: !(*opts.DisableAutomaticPunctuation),
					NumChannels:                uint32(*opts.NumChannels),
				},
				SingleUtterance: *opts.SingleUtterance,
				InterimResultsConfig: &sttPb.InterimResultsConfig{
					EnableInterimResults: *opts.InterimResults,
				},
			},
		},
	}
	if *opts.DoNotPerformVad {
		reqeust.GetStreamingConfig().Config.Vad = &sttPb.RecognitionConfig_DoNotPerformVad{DoNotPerformVad: true}
	} else {
		reqeust.GetStreamingConfig().Config.Vad = &sttPb.RecognitionConfig_VadConfig{
			VadConfig: &sttPb.VoiceActivityDetectionConfig{
				SilenceDurationThreshold: float32(*opts.SilenceDurationThreshold),
			},
		}
	}

	if err := stream.Send(reqeust); err != nil {
		panic(err)
	}
	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		defer stream.CloseSend()
		for {
			buffer := make([]byte, 1024)
			bytesRead, err := dataReader.Read(buffer)

			if err == io.EOF {
				break
			}
			if err != nil {
				panic(err)
			}

			err = stream.Send(&sttPb.StreamingRecognizeRequest{
				StreamingRequest: &sttPb.StreamingRecognizeRequest_AudioContent{
					AudioContent: buffer[:bytesRead],
				},
			})
			if err != nil {
				panic(err)
			}
		}

	}()

	for {
		msg, err := stream.Recv()
		if err == io.EOF {
			break
		}
		if err != nil {
			panic(err)
		}
		if common.PrettyPrintProtobuf(msg) != nil {
			panic(err)
		}
	}

	wg.Wait()

}
